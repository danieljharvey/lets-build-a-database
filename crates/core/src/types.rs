use std::{fmt::Display, hash::Hash};

#[derive(Debug, PartialOrd, PartialEq, Eq, Ord, Hash, Clone)]
pub struct Column {
    pub name: String,
    pub table_alias: Option<TableAlias>,
}

impl Display for Column {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match &self.table_alias {
            Some(table_alias) => {
                write!(f, "{}.{}", table_alias, self.name)
            }
            None => {
                write!(f, "{}", self.name)
            }
        }
    }
}

impl std::convert::From<&str> for Column {
    fn from(name: &str) -> Column {
        Column {
            name: name.to_string(),
            table_alias: None,
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum Expr {
    ColumnComparison {
        column: Column,
        op: Op,
        literal: serde_json::Value,
    },
}

#[derive(Debug, PartialEq)]
pub enum Op {
    Equals,
    GreaterThan,
    GreaterThanOrEqual,
    LessThan,
    LessThanOrEqual,
}

#[derive(Debug, PartialEq)]
pub struct Join {
    #[allow(clippy::struct_field_names)]
    pub join_type: JoinType,
    pub left_from: Box<Query>,
    pub right_from: Box<Query>,
    pub left_column_on: Column,
    pub right_column_on: Column,
}

#[derive(Debug, PartialEq)]
pub struct TableName(pub String);

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
pub struct TableAlias(pub String);

impl Display for TableAlias {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        write!(f, "{}", self.0)
    }
}

#[derive(Debug, PartialEq)]
pub struct From {
    pub table_name: TableName,
    pub table_alias: Option<TableAlias>,
}

#[derive(Debug, PartialEq)]
pub struct Filter {
    pub from: Box<Query>,
    pub filter: Expr,
}

#[derive(Debug, PartialEq)]
pub struct Project {
    pub from: Box<Query>,
    pub fields: Vec<Column>,
}

#[derive(Debug, PartialEq)]
pub enum JoinType {
    Inner,
    LeftOuter,
}

#[derive(Debug, PartialEq)]
pub struct Limit {
    pub from: Box<Query>,
    pub limit: u64,
}

#[derive(Debug, PartialEq)]
pub enum Query {
    From(From),
    Filter(Filter),
    Join(Join),
    Project(Project),
    Limit(Limit),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Row {
    pub items: Vec<serde_json::Value>,
}

impl Row {
    pub fn get_column(&self, column: &Column, schema: &Schema) -> Option<&serde_json::Value> {
        let index = schema.get_index_for_column(column)?;

        self.items.get(index)
    }

    pub fn extend(&mut self, row: Row) {
        self.items.extend(row.items);
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct Schema {
    pub columns: Vec<Column>,
}

impl Schema {
    pub fn get_index_for_column(&self, column: &Column) -> Option<usize> {
        self.columns
            .iter()
            .enumerate()
            .find(|(_, column_name)| column_name == &column)
            .map(|(i, _)| i)
    }

    pub fn extend(&mut self, schema: Schema) {
        self.columns.extend(schema.columns);
    }
}

pub struct QueryStep {
    pub schema: Schema,
    pub rows: Vec<Row>,
}

impl QueryStep {
    // reconstruct JSON output
    pub fn to_json(&self) -> serde_json::Value {
        let mut output_rows = vec![];

        for row in &self.rows {
            let mut output_row = serde_json::Map::new();
            for column in &self.schema.columns {
                let value = row.get_column(column, &self.schema).unwrap();
                output_row.insert(column.to_string(), value.clone());
            }
            output_rows.push(serde_json::Value::Object(output_row));
        }
        serde_json::Value::Array(output_rows)
    }
}
