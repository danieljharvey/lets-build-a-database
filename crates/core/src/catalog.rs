use std::collections::BTreeMap;

use crate::{
    indexes::{construct_index, ConstructedIndexes, Index},
    query::raw_rows_for_table,
    types::{ColumnName, TableName},
};

pub struct Table {
    pub columns: Vec<ColumnName>,
    pub indexes: Vec<Index>,
}

pub struct Catalog {
    pub tables: BTreeMap<TableName, Table>,
}

impl Catalog {
    pub fn construct_indexes(&self) -> ConstructedIndexes {
        let mut constructed_indexes = BTreeMap::new();
        for (table_name, table) in &self.tables {
            let mut indexes = vec![];
            for index in &table.indexes {
                let rows = raw_rows_for_table(&index.table_name);
                indexes.push((index.clone(), construct_index(index, &rows)))
            }
            constructed_indexes.insert(table_name.clone(), indexes);
        }
        ConstructedIndexes {
            indexes: constructed_indexes,
        }
    }
}

// columns and indexes for each table
pub fn get_static_catalog() -> Catalog {
    let animal = Table {
        columns: vec![
            "animal_id".into(),
            "animal_name".into(),
            "species_id".into(),
        ],
        indexes: vec![
            Index {
                table_name: "animal".into(),
                columns: vec!["animal_id".into()],
            },
            Index {
                table_name: "animal".into(),
                columns: vec!["species_id".into()],
            },
        ],
    };

    let species = Table {
        columns: vec!["species_id".into(), "species_name".into()],
        indexes: vec![Index {
            table_name: "species".into(),
            columns: vec!["species_id".into()],
        }],
    };

    let album = Table {
        columns: vec!["AlbumId".into(), "Title".into(), "ArtistId".into()],
        indexes: vec![
            Index {
                table_name: "Album".into(),
                columns: vec!["AlbumId".into()],
            },
            Index {
                table_name: "Album".into(),
                columns: vec!["ArtistId".into()],
            },
        ],
    };

    let artist = Table {
        columns: vec!["ArtistId".into(), "Name".into()],
        indexes: vec![Index {
            table_name: "Artist".into(),
            columns: vec!["ArtistId".into()],
        }],
    };

    let track = Table {
        columns: vec![
            "TrackId".into(),
            "Name".into(),
            "AlbumId".into(),
            "MediaTypeId".into(),
            "GenreId".into(),
            "Composer".into(),
            "Milliseconds".into(),
            "Bytes".into(),
            "UnitPrice".into(),
        ],
        indexes: vec![
            Index {
                table_name: "Track".into(),
                columns: vec!["TrackId".into()],
            },
            Index {
                table_name: "Track".into(),
                columns: vec!["AlbumId".into()],
            },
            Index {
                table_name: "Track".into(),
                columns: vec!["MediaTypeId".into()],
            },
            Index {
                table_name: "Track".into(),
                columns: vec!["GenreId".into()],
            },
        ],
    };

    Catalog {
        tables: BTreeMap::from_iter([
            ("animal".into(), animal),
            ("species".into(), species),
            ("Album".into(), album),
            ("Artist".into(), artist),
            ("Track".into(), track),
        ]),
    }
}
