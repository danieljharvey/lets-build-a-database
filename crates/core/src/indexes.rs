use std::collections::BTreeMap;
use std::hash::DefaultHasher;
use std::hash::Hash;
use std::hash::Hasher;

use crate::types::{ColumnName, TableName};

#[derive(Debug, PartialEq)]
pub struct LineNumber(pub usize);

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct HashedValue(pub u64);

// the index we want
#[derive(Debug, PartialEq, Clone)]
pub struct Index {
    pub table_name: TableName,
    pub columns: Vec<ColumnName>,
}

// the items
#[derive(Debug, PartialEq)]
pub struct ConstructedIndex {
    items: BTreeMap<HashedValue, Vec<LineNumber>>,
}

pub fn construct_index(index: &Index, rows: &[serde_json::Value]) -> ConstructedIndex {
    let mut items: BTreeMap<HashedValue, Vec<LineNumber>> = BTreeMap::new();

    for (i, row) in rows.iter().enumerate() {
        let mut values = vec![];
        for column in &index.columns {
            if let Some(value) = row.get(column.0.as_str()) {
                values.push(value.clone());
            } else {
                values.push(serde_json::Value::Null);
            }
        }

        let hash = HashedValue(calculate_hash(&serde_json::Value::Array(values)));

        if let Some(line_numbers) = items.get_mut(&hash) {
            // add another item
            line_numbers.push(LineNumber(i));
        } else {
            items.insert(hash, vec![LineNumber(i)]);
        }
    }
    ConstructedIndex { items }
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use serde_json::json;

    use crate::{
        indexes::{
            calculate_hash, construct_index, ConstructedIndex, HashedValue, Index, LineNumber,
        },
        types::{ColumnName, TableName},
    };

    #[test]
    fn test_construct_index() {
        let rows = vec![
            json!({"id": 1}),
            json!({"id": 2}),
            json!({"id": 3}),
            json!({"id": 3}),
        ];

        let index = Index {
            table_name: TableName("test".into()),
            columns: vec![ColumnName("id".into())],
        };

        let one_hash = HashedValue(calculate_hash(&json!([1])));
        let two_hash = HashedValue(calculate_hash(&json!([2])));
        let three_hash = HashedValue(calculate_hash(&json!([3])));

        let expected = ConstructedIndex {
            items: BTreeMap::from_iter([
                (one_hash, vec![LineNumber(0)]),
                (two_hash, vec![LineNumber(1)]),
                (three_hash, vec![LineNumber(2), LineNumber(3)]),
            ]),
        };

        assert_eq!(construct_index(&index, &rows), expected)
    }
}

fn calculate_hash<T: Hash>(t: &T) -> u64 {
    let mut s = DefaultHasher::new();
    t.hash(&mut s);
    s.finish()
}
