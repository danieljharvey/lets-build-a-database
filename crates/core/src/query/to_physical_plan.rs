use std::collections::BTreeMap;

use crate::indexes::{ConstructedIndexes, Index};
use crate::types::{ColumnName, Expr, IndexScan, Limit, Op, PhysicalPlan, TableName, TableScan};

use crate::types::{Filter, From, Join, LogicalPlan, Project};

pub fn to_physical_plan(logical_plan: LogicalPlan, indexes: &ConstructedIndexes) -> PhysicalPlan {
    match logical_plan {
        LogicalPlan::From(From {
            table_name,
            table_alias,
        }) => PhysicalPlan::TableScan(TableScan {
            table_name,
            table_alias,
        }),
        LogicalPlan::Filter(filter) => filter_to_physical_plan(filter, indexes),
        LogicalPlan::Join(Join {
            join_type,
            left_from,
            right_from,
            on,
        }) => PhysicalPlan::Join(Join {
            join_type,
            on,
            left_from: Box::new(to_physical_plan(*left_from, indexes)),
            right_from: Box::new(to_physical_plan(*right_from, indexes)),
        }),
        LogicalPlan::Limit(Limit { from, limit }) => PhysicalPlan::Limit(Limit {
            from: Box::new(to_physical_plan(*from, indexes)),
            limit,
        }),
        LogicalPlan::Project(Project { from, fields }) => PhysicalPlan::Project(Project {
            from: Box::new(to_physical_plan(*from, indexes)),
            fields,
        }),
    }
}

fn columns_in_filter(expr: &Expr) -> BTreeMap<&ColumnName, FilterValue> {
    match expr {
        Expr::ColumnComparison {
            column,
            literal,
            op: Op::Equals,
        } => BTreeMap::from_iter([(&column.name, FilterValue::Eq(literal))]),
        Expr::ColumnComparison { .. } => BTreeMap::new(),
    }
}

enum FilterValue<'a> {
    Eq(&'a serde_json::Value),
}

// convert filter to physical plan. it may contain a `From` inside, in which case combine them
// into an IndexScan
fn filter_to_physical_plan(
    filter: Filter<LogicalPlan>,
    indexes: &ConstructedIndexes,
) -> PhysicalPlan {
    let Filter { from, filter } = filter;

    let inner = match from.as_ref() {
        LogicalPlan::From(From {
            table_name,
            table_alias,
        }) => {
            let filter_columns = columns_in_filter(&filter);
            if let Some((index, values)) = find_index(indexes, table_name, &filter_columns) {
                // when we have multiple comparisons in a filter,
                // remove column from filter (later)
                // for now, remove filter altogether
                // create index scan
                return PhysicalPlan::IndexScan(IndexScan {
                    index: index.clone(),
                    table_name: table_name.clone(),
                    table_alias: table_alias.clone(),
                    values,
                });
            }
            to_physical_plan(*from, indexes)
        }
        _ => to_physical_plan(*from, indexes),
    };

    PhysicalPlan::Filter(Filter {
        from: Box::new(inner),
        filter,
    })
}

fn find_index<'index>(
    indexes: &'index ConstructedIndexes,
    table_name: &'_ TableName,
    filter_columns: &BTreeMap<&'_ ColumnName, FilterValue<'_>>,
) -> Option<(&'index Index, Vec<serde_json::Value>)> {
    if let Some(indexes_for_table) = indexes.indexes.get(table_name) {
        for (index, _) in indexes_for_table {
            if let Some(values) = index
                .columns
                .iter()
                .map(|column| {
                    filter_columns.get(column).map(|val| match val {
                        FilterValue::Eq(val) => val,
                    })
                })
                .collect::<Option<Vec<_>>>()
            {
                // there's only one value here for now
                let json_value =
                    serde_json::Value::Array(values.into_iter().copied().cloned().collect());

                return Some((index, vec![json_value]));
            }
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use serde_json::json;

    use crate::{
        indexes::{construct_index, ConstructedIndexes, Index},
        query::{from::raw_rows_for_table, to_physical_plan::to_physical_plan},
        types::{Expr, Filter, From, IndexScan, LogicalPlan, Op, PhysicalPlan},
    };

    #[test]
    fn test_filter_to_index_scan() {
        let artist_primary_index = Index {
            table_name: "Artist".into(),
            columns: vec!["ArtistId".into()],
        };

        let rows = raw_rows_for_table(&"Artist".into());

        let constructed_artist_primary_index = construct_index(&artist_primary_index, &rows);

        let indexes: ConstructedIndexes = ConstructedIndexes {
            indexes: BTreeMap::from_iter(
                vec![(
                    "Artist".into(),
                    vec![(
                        artist_primary_index.clone(),
                        constructed_artist_primary_index,
                    )],
                )]
                .into_iter(),
            ),
        };

        let logical_plan = LogicalPlan::Filter(Filter {
            from: Box::new(LogicalPlan::From(From {
                table_name: "Artist".into(),
                table_alias: None,
            })),
            filter: Expr::ColumnComparison {
                column: "ArtistId".into(),
                op: Op::Equals,
                literal: 1.into(),
            },
        });

        let expected = PhysicalPlan::IndexScan(IndexScan {
            table_name: "Artist".into(),
            table_alias: None,
            index: artist_primary_index,
            values: vec![json!([1])],
        });

        assert_eq!(to_physical_plan(logical_plan, &indexes), expected);
    }
}
