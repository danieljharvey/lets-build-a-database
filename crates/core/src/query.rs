mod filter;
mod from;
mod join;
mod project;
mod to_physical_plan;

pub use from::raw_rows_for_table;

use crate::catalog::{self, Catalog};
use crate::indexes::ConstructedIndexes;
use crate::types::{IndexScan, Limit, PhysicalPlan, TableScan};
use to_physical_plan::to_physical_plan;

use super::types::QueryStep;
use super::types::{Column, Filter, Join, LogicalPlan, Project};

#[derive(Debug)]
pub enum QueryError {
    ColumnNotFoundInSchema { column_name: Column },
    IndexNotFoundInSchema { index: usize },
    FilterError(filter::FilterError),
}

fn run_physical_plan(
    physical_plan: &PhysicalPlan,
    catalog: &Catalog,
) -> Result<QueryStep, QueryError> {
    match physical_plan {
        PhysicalPlan::TableScan(TableScan {
            table_name,
            table_alias,
        }) => Ok(from::table_scan(table_name, table_alias.as_ref(), catalog)),
        PhysicalPlan::IndexScan(IndexScan {
            table_name,
            table_alias,
            index,
            values,
        }) => Ok(from::index_scan(
            table_name,
            table_alias.as_ref(),
            index,
            values.catalog,
        )),
        PhysicalPlan::Filter(Filter { from, filter }) => {
            let QueryStep {
                schema,
                rows,
                mut cost,
            } = run_physical_plan(from, catalog)?;

            let mut filtered_rows = vec![];

            for row in rows {
                cost.increment_rows_processed();
                if filter::apply_predicate(&row, &schema, filter)? {
                    filtered_rows.push(row);
                }
            }

            Ok(QueryStep {
                schema,
                rows: filtered_rows,
                cost,
            })
        }
        PhysicalPlan::Project(Project { from, fields }) => {
            let QueryStep {
                schema,
                rows,
                mut cost,
            } = run_physical_plan(from, catalog)?;

            let mut projected_rows = vec![];

            for row in &rows {
                cost.increment_rows_processed();
                projected_rows.push(project::project_fields(row, &schema, fields)?);
            }

            let schema = project::project_schema(&schema, fields)?;

            Ok(QueryStep {
                schema,
                rows: projected_rows,
                cost,
            })
        }
        PhysicalPlan::Limit(Limit { limit, from }) => {
            let QueryStep {
                schema,
                mut rows,
                cost,
            } = run_physical_plan(from, catalog)?;
            let size: usize = (*limit).try_into().unwrap();

            rows.truncate(size);

            Ok(QueryStep { schema, rows, cost })
        }
        PhysicalPlan::Join(Join {
            left_from,
            right_from,
            join_type,
            on,
        }) => {
            let QueryStep {
                schema: left_schema,
                rows: left_rows,
                cost: mut left_cost,
            } = run_physical_plan(left_from, catalog)?;

            let QueryStep {
                schema: right_schema,
                rows: right_rows,
                cost: right_cost,
            } = run_physical_plan(right_from, catalog)?;

            left_cost.extend(&right_cost);

            join::hash_join(
                left_rows,
                &left_schema,
                right_rows,
                &right_schema,
                on,
                join_type,
                left_cost,
            )
        }
    }
}

// TODO: pre-calculate indexes and pass them in
pub fn run_query(
    query: LogicalPlan,
    constructed_indexes: &ConstructedIndexes,
    catalog: &Catalog,
) -> Result<QueryStep, QueryError> {
    let physical_plan = to_physical_plan(query, constructed_indexes);
    run_physical_plan(&physical_plan, catalog)
}

#[cfg(test)]
mod tests {
    use crate::{
        catalog::get_static_catalog,
        parser::parse,
        run_query,
        types::{LogicalPlan, QueryStep},
    };

    use super::QueryError;

    fn test_run_query(query: LogicalPlan) -> Result<QueryStep, QueryError> {
        let catalog = get_static_catalog();
        let constructed_indexes = catalog.construct_indexes();
        run_query(query, &constructed_indexes, &catalog)
    }

    #[test]
    fn test_query_select_animals() {
        let query = parse("SELECT * FROM animal").unwrap();

        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_query_select_one_animal() {
        let query = parse("SELECT * FROM animal where animal_id = 1").unwrap();

        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_query_select_horse() {
        let query = parse("select * from animal where animal_name = 'horse'").unwrap();
        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_query_projection() {
        let query = parse("select animal_name from animal where animal_name = 'horse'").unwrap();
        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_horse_and_species() {
        let query = parse(
            r#"
        select * from animal 
        join species 
            on species_id 
        where animal_name = 'horse'"#,
        )
        .unwrap();
        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_species_and_animals() {
        let query = parse(
            r#"
        select * from species 
          join animal on species_id
        where
          species_id >= 3 
    "#,
        )
        .unwrap();
        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_species_and_animals_left_outer() {
        let query = parse(
            r#"
        select * from species 
          left outer join animal on species_id
        where
          species_id = 3 
    "#,
        )
        .unwrap();
        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_album() {
        let query = parse(
            r#"
        select * from Album 
        where Title = 'Jagged Little Pill'
    "#,
        )
        .unwrap();
        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_album_and_artist() {
        let query = parse(
            r#"
        select * from Album 
          join Artist on ArtistId
        where
          ArtistId = 82
    "#,
        )
        .unwrap();
        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_track_album_and_artist() {
        let query = parse(
            r#"
        select Name, Title, artist.Name from Track
          join Album on AlbumId
          join Artist as artist on ArtistId
        where
          ArtistId = 82
        limit 10
    "#,
        )
        .unwrap();
        let result = test_run_query(query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }
}
