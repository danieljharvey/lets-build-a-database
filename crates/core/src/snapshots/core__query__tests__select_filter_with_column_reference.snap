---
source: crates/core/src/query.rs
expression: result.to_json()
---
[
  {
    "AlbumId": 3,
    "ArtistId": 2,
    "Title": "Restless and Wild"
  },
  {
    "AlbumId": 59,
    "ArtistId": 58,
    "Title": "Deep Purple In Rock"
  },
  {
    "AlbumId": 88,
    "ArtistId": 87,
    "Title": "Faceless"
  }
]
